File: src/org/wavescale/sourcesync/synchronizer/SFTPFileSynchronizer.java
package org.wavescale.sourcesync.synchronizer;

import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.project.Project;
import com.jcraft.jsch.*;
import org.jetbrains.annotations.NotNull;
import org.wavescale.sourcesync.api.FileSynchronizer;
import org.wavescale.sourcesync.api.Utils;
import org.wavescale.sourcesync.config.SFTPConfiguration;
import org.wavescale.sourcesync.logger.EventDataLogger;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * ****************************************************************************
 * Copyright (c) 2014-2107 Faur Ioan-Aurel.                                     *
 * All rights reserved. This program and the accompanying materials             *
 * are made available under the terms of the MIT License                        *
 * which accompanies this distribution, and is available at                     *
 * http://opensource.org/licenses/MIT                                           *
 * *
 * For any issues or questions send an email at: fioan89@gmail.com              *
 * *****************************************************************************
 */
public class SFTPFileSynchronizer extends FileSynchronizer {

    public static final String SSH_KNOWN_HOSTS = Paths.get(System.getProperty("user.home"), ".ssh", "known_hosts").toString();
    private final JSch jsch;
    private Session session;

    /**
     * Build a file synchronizer from general info contained by <b>connectionInfo</b> param.
     *
     * @param connectionInfo a {@link org.wavescale.sourcesync.config.SFTPConfiguration} instance
     *                       containing session info like hostname, user, password, etc...
     * @param project        a {@link com.intellij.openapi.project.Project} instance used to gather project relative
     *                       metadata like name, absoulte path, etc...
     * @param indicator      used to report progress on upload process.
     */
    public SFTPFileSynchronizer(@NotNull SFTPConfiguration connectionInfo, @NotNull Project project, @NotNull ProgressIndicator indicator) {
        super(connectionInfo, project, indicator);
        this.jsch = new JSch();
        this.getIndicator().setIndeterminate(true);
    }

    @Override
    public boolean connect() {
        if (!isConnected()) {
            try {
                initSession();
                session.setPassword(this.getConnectionInfo().getUserPassword());
                this.session.setConfig("StrictHostKeyChecking", "no");
                this.session.connect();
                this.setConnected(true);
                return true;
            } catch (JSchException e) {
                EventDataLogger.logWarning(e.toString(), this.getProject());
                return false;
            }
        }
        return true;
    }

    private void initSession() throws JSchException {
        SFTPConfiguration configuration = (SFTPConfiguration) this.getConnectionInfo();
        session = this.jsch.getSession(this.getConnectionInfo().getUserName(), this.getConnectionInfo().getHost(),
                this.getConnectionInfo().getPort());
        if (configuration.isPasswordlessSSHSelected()) {
            session.setConfig("PreferredAuthentications", "publickey");
            try {
                Utils.createFile(SSH_KNOWN_HOSTS);
            } catch (IOException e) {
                EventDataLogger.logError("Could not identify nor create the ssh known hosts file at " + SSH_KNOWN_HOSTS + ". The returned error is:" + e.getMessage(), this.getProject());
            }
            this.jsch.setKnownHosts(SSH_KNOWN_HOSTS);
            // add private key and passphrase if exists
            if (configuration.isPasswordlessWithPassphrase()) {
                this.jsch.addIdentity(configuration.getCertificatePath(), configuration.getUserPassword());
            } else {
                this.jsch.addIdentity(configuration.getCertificatePath());
            }

        } else {
            session.setPassword(this.getConnectionInfo().getUserPassword());
        }
    }

    @Override
    public void disconnect() {
        if (this.session != null) {
            this.session.disconnect();
            this.setConnected(false);
        }
    }

    @Override
    public void syncFile(String sourcePath, Path uploadLocation) {
        boolean preserveTimestamp = this.getConnectionInfo().isPreserveTime();
        Path remotePath = Paths.get(this.getConnectionInfo().getRootPath()).resolve(uploadLocation);

        ChannelSftp channelSftp;
        try {
            channelSftp = (ChannelSftp) this.session.openChannel("sftp");
            channelSftp.connect();
        } catch (JSchException e) {
            EventDataLogger.logError(e.toString(), this.getProject());
            return;
        }

        // first try to create the path where this must be uploaded
        try {
            channelSftp.cd("/");
        } catch (SftpException e) {
            EventDataLogger.logError("On remote we could not change directory into root: " + remotePath.getRoot(), this.getProject());
        }
        for (Path current : remotePath) {
            String location = current.toString();
            try {
                channelSftp.mkdir(location);
            } catch (SftpException e) {
                // this dir probably exist so just ignore
            }
            try {
                channelSftp.cd(location);
            } catch (SftpException e) {
                // probably it doesn't exist or maybe no permission
                EventDataLogger.logError("Remote dir <b>" + remotePath +
                        "</b> might not exist or you don't have permission on this path!", this.getProject());
                return;
            }
        }

        // upload file
        File toUpload = new File(sourcePath);
        SftpProgressMonitor progressMonitor = new SftpMonitor(toUpload.length());
        try {
            channelSftp.put(new FileInputStream(toUpload), toUpload.getName(), progressMonitor, ChannelSftp.OVERWRITE);
            if (preserveTimestamp) {
                SftpATTRS sftpATTRS = channelSftp.lstat(toUpload.getName());
                int lastAcc = sftpATTRS.getATime();
                // this is a messed method: if lastModified is greater than Integer.MAX_VALUE
                // then timestamp will not be ok.
                sftpATTRS.setACMODTIME(lastAcc, new Long(toUpload.lastModified() / 1000).intValue());
                channelSftp.setStat(toUpload.getName(), sftpATTRS);
            }
        } catch (SftpException e) {
            EventDataLogger.logWarning(e.toString(), getProject());
        } catch (FileNotFoundException e) {
            EventDataLogger.logWarning(e.toString(), getProject());
        }

        channelSftp.disconnect();
    }

    private class SftpMonitor implements SftpProgressMonitor {
        final double totalLength;
        long totalUploaded;

        public SftpMonitor(long totalLength) {
            this.totalLength = totalLength + 0.0;
            this.totalUploaded = 0;
        }

        @Override
        public void init(int opcode, String src, String dest, long max) {
            File remoteFile = new File(dest);
            if (SftpProgressMonitor.PUT == opcode) {
                SFTPFileSynchronizer.this.getIndicator().setText("Uploading...[" + remoteFile.getName() + "]");
                SFTPFileSynchronizer.this.getIndicator().setIndeterminate(false);

            }

        }

        @Override
        public boolean count(long count) {
            totalUploaded += count;
            SFTPFileSynchronizer.this.getIndicator().setFraction(totalUploaded / totalLength);
            // false will kill the upload
            return true;
        }

        @Override
        public void end() {
            SFTPFileSynchronizer.this.getIndicator().setFraction(1.0);
        }
    }
}

File: src/main/kotlin/org/wavescale/sourcesync/synchronizer/SFTPFileSynchronizer.kt
/**
 * ****************************************************************************
 * Copyright (c) 2014-2022 Faur Ioan-Aurel.                                     *
 * All rights reserved. This program and the accompanying materials             *
 * are made available under the terms of the MIT License                        *
 * which accompanies this distribution, and is available at                     *
 * http://opensource.org/licenses/MIT                                           *
 * *
 * For any issues or questions send an email at: fioan89@gmail.com              *
 * *****************************************************************************
 */
package org.wavescale.sourcesync.synchronizer

import com.intellij.openapi.components.service
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.Project
import com.jcraft.jsch.ChannelSftp
import com.jcraft.jsch.JSch
import com.jcraft.jsch.JSchException
import com.jcraft.jsch.Session
import com.jcraft.jsch.SftpProgressMonitor
import org.wavescale.sourcesync.SourcesyncBundle
import org.wavescale.sourcesync.api.Utils
import org.wavescale.sourcesync.configurations.AuthenticationType
import org.wavescale.sourcesync.configurations.SshSyncConfiguration
import org.wavescale.sourcesync.notifications.Notifier
import org.wavescale.sourcesync.services.StatsService
import org.wavescale.sourcesync.services.SyncStatusService
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.nio.file.Path
import java.nio.file.Paths


/**
 * Build a file synchronizer from general info contained by **configuration** param.
 */
class SFTPFileSynchronizer(private val configuration: SshSyncConfiguration, val project: Project) : Synchronizer {
    private val syncStatusService = service<SyncStatusService>()
    private val statsService = service<StatsService>()

    private val jsch: JSch = JSch()
    private var session: Session? = null

    private var isConnected: Boolean = false
    override fun connect(): Boolean {
        return if (!isConnected) {
            try {
                initSession()
                session!!.connect()
                isConnected = true
                true
            } catch (e: JSchException) {
                syncStatusService.removeRunningSync(configuration.name)
                Notifier.notifyError(
                    project,
                    SourcesyncBundle.message("ssh.upload.fail.text"),
                    "Can't open SSH connection to ${configuration.hostname}. Reason: ${e.message}",
                )
                false
            }
        } else true
    }

    @Throws(JSchException::class)
    private fun initSession() {
        syncStatusService.addRunningSync(configuration.name)
        session = jsch.getSession(configuration.username, configuration.hostname, configuration.port.toInt())
        session!!.setConfig("StrictHostKeyChecking", "no")
        if (configuration.authenticationType == AuthenticationType.KEY_PAIR) {
            session!!.setConfig("PreferredAuthentications", "publickey")
            try {
                Utils.createFile(SSH_KNOWN_HOSTS)
            } catch (e: IOException) {
                syncStatusService.removeRunningSync(configuration.name)
                Notifier.notifyError(
                    project,
                    SourcesyncBundle.message("ssh.upload.fail.text"),
                    "Could not identify nor create the SSH known hosts file at ${SCPFileSynchronizer.SSH_KNOWN_HOSTS}. Reason: ${e.message}",
                )
                return
            }
            jsch.setKnownHosts(SSH_KNOWN_HOSTS)
            // add private key and passphrase if exists
            if (configuration.passphrase.isNullOrEmpty().not()) {
                jsch.addIdentity(configuration.privateKey, configuration.passphrase)
            } else {
                jsch.addIdentity(configuration.privateKey)
            }
        } else {
            session!!.setPassword(configuration.password)
        }
    }

    override fun disconnect() {
        try {
            session?.disconnect()
        } finally {
            isConnected = false
            syncStatusService.removeRunningSync(configuration.name)
        }
    }

    override fun syncFile(sourcePath: String, uploadLocation: Path, indicator: ProgressIndicator) {
        val preserveTimestamp = configuration.preserveTimestamps
        val channelSftp: ChannelSftp
        try {
            channelSftp = session!!.openChannel("sftp") as ChannelSftp
            channelSftp.connect()
        } catch (e: JSchException) {
            syncStatusService.removeRunningSync(configuration.name)
            Notifier.notifyError(
                project,
                SourcesyncBundle.message("ssh.upload.fail.text"),
                "An error was encountered while trying to open a SSH connection to ${configuration.hostname}. Reason: ${e.message}",
            )
            return
        }

        if (!channelSftp.absoluteDirExists(configuration.workspaceBasePath)) {
            syncStatusService.removeRunningSync(configuration.name)
            Notifier.notifyError(
                project,
                SourcesyncBundle.message("ssh.upload.fail.text"),
                "Remote project base path ${configuration.workspaceBasePath} does not exist or is not a directory. Please make sure the value is a valid absolute directory path on ${configuration.hostname}",
            )
            return
        }

        channelSftp.cd(configuration.workspaceBasePath)

        if (!channelSftp.localDirExistsOnRemote(uploadLocation.toString())) {
            logger.info("Upload path $uploadLocation does not exist or is not a directory. Going to create it.")
            val exists = channelSftp.mkLocalDirsOnRemote(uploadLocation.toString())
            if (!exists) {
                syncStatusService.removeRunningSync(configuration.name)
                Notifier.notifyError(
                    project,
                    SourcesyncBundle.message("ssh.upload.fail.text"),
                    "Upload path $uploadLocation could not be created on ${configuration.hostname}"
                )
                return
            }
        }
        if (!channelSftp.cdLocalDirsOnRemote(uploadLocation.toString())) {
            syncStatusService.removeRunningSync(configuration.name)
            Notifier.notifyError(
                project,
                SourcesyncBundle.message("ssh.upload.fail.text"),
                "Could not change directory to $uploadLocation"
            )
            return
        }

        // upload file
        val toUpload = File(sourcePath)
        val progressMonitor: SftpProgressMonitor = SftpMonitor(toUpload.length(), indicator)
        try {
            channelSftp.put(FileInputStream(toUpload), toUpload.name, progressMonitor, ChannelSftp.OVERWRITE)
            if (preserveTimestamp) {
                val sftpATTRS = channelSftp.lstat(toUpload.name)
                val lastAcc = sftpATTRS.aTime
                // this is a messed method: if lastModified is greater than Integer.MAX_VALUE
                // then timestamp will not be ok.
                sftpATTRS.setACMODTIME(lastAcc, java.lang.Long.valueOf(toUpload.lastModified() / 1000).toInt())
                channelSftp.setStat(toUpload.name, sftpATTRS)
            }
        } catch (e: Exception) {
            syncStatusService.removeRunningSync(configuration.name)
            Notifier.notifyError(
                project,
                SourcesyncBundle.message("ssh.upload.fail.text"),
                "Upload to ${configuration.hostname} failed. Reason: ${e.message}"
            )
        }
        channelSftp.disconnect()
    }

    private inner class SftpMonitor(totalLength: Long, private val indicator: ProgressIndicator) : SftpProgressMonitor {
        val totalLength: Double
        var totalUploaded: Long

        init {
            this.totalLength = totalLength + 0.0
            totalUploaded = 0
        }

        override fun init(opcode: Int, src: String, dest: String, max: Long) {
            val remoteFile = File(dest)
            if (SftpProgressMonitor.PUT == opcode) {
                indicator.text = "Uploading...[" + remoteFile.name + "]"
                indicator.isIndeterminate = false
            }
        }

        override fun count(count: Long): Boolean {
            totalUploaded += count
            indicator.fraction = totalUploaded / totalLength
            // false will kill the upload
            return true
        }

        override fun end() {
            indicator.fraction = 1.0
            statsService.registerSuccessfulUpload()
            if (statsService.eligibleForDonations()) {
                Notifier.notifyDonation(project)
            }
        }
    }

    companion object {
        val SSH_KNOWN_HOSTS = Paths.get(System.getProperty("user.home"), ".ssh", "known_hosts").toString()

        private val logger = logger<SFTPFileSynchronizer>()
    }
}